source/common/network/connection_impl.cc

```diff
@@ -349,11 +350,14 @@ void ConnectionImpl::readDisable(bool disable) {
       file_event_->setEnabled(Event::FileReadyType::Read | Event::FileReadyType::Write);
     }
 
-    if (consumerWantsToRead() && read_buffer_.length() > 0) {
-      // If the connection has data buffered there's no guarantee there's also data in the kernel
-      // which will kick off the filter chain. Alternately if the read buffer has data the fd could
-      // be read disabled. To handle these cases, fake an event to make sure the buffered data gets
-      // processed regardless and ensure that we dispatch it via onRead.
+    if ((consumerWantsToRead() && read_buffer_.length() > 0) ||
+        (read_disable_count_ == 0 && want_read_)) {
```

Comment:
Sorry one other question from another look: is it possible to somehow fold all of this logic into `consumerWantsToRead()`? It seems kind of odd that we have a function that is tracking the consumer wants to read and then a different wants to read boolean. If it's too hard to merge them maybe some better names for both the function and the boolean?

Reply:
Tried to improve on names and the if predicate.